File: src/position-manager.js
/**
 * Position Manager - Manages trading positions with stop-loss and take-profit
 * Based on CloddsBot positions skill documentation
 * Source: tessl.io/registry/skills/github/alsk1992/CloddsBot/positions
 */

import EventEmitter from 'events';

class PositionManager extends EventEmitter {
  constructor(options = {}) {
    super();
    this.positions = new Map();
    this.checkIntervalMs = options.checkIntervalMs || 5000;
    this.orderType = options.orderType || 'market';
    this.limitBuffer = options.limitBuffer || 0.01;
    this.intervalId = null;
  }

  /**
   * Start monitoring positions
   */
  async start() {
    if (this.intervalId) return;
    
    this.intervalId = setInterval(() => {
      this.monitorPositions();
    }, this.checkIntervalMs);
    
    console.log('Position manager started');
  }

  /**
   * Stop monitoring
   */
  stop() {
    if (this.intervalId) {
      clearInterval(this.intervalId);
      this.intervalId = null;
    }
  }

  /**
   * Add or update a position
   * @param {Object} position - Position data
   */
  addPosition(position) {
    const pos = {
      id: position.id || this.generateId(),
      platform: position.platform,
      market: position.market,
      side: position.side,
      size: position.size,
      entryPrice: position.entryPrice,
      currentPrice: position.entryPrice,
      stopLoss: position.stopLoss || null,
      takeProfit: position.takeProfit || null,
      trailingStop: position.trailingStop || null,
      highWaterMark: position.entryPrice,
      createdAt: new Date(),
      updatedAt: new Date()
    };
    
    this.positions.set(pos.id, pos);
    this.emit('positionAdded', pos);
    
    return pos;
  }

  /**
   * Update position price
   * @param {string} positionId - Position identifier
   * @param {number} currentPrice - Current market price
   */
  updatePrice(positionId, currentPrice) {
    const pos = this.positions.get(positionId);
    if (!pos) return null;
    
    pos.currentPrice = currentPrice;
    pos.updatedAt = new Date();
    
    // Update high water mark for trailing stop
    if (currentPrice > pos.highWaterMark) {
      pos.highWaterMark = currentPrice;
    }
    
    this.positions.set(positionId, pos);
    return pos;
  }

  /**
   * List all positions
   * @param {string} platform - Filter by platform (optional)
   */
  listPositions(platform = null) {
    const positions = Array.from(this.positions.values());
    
    if (platform) {
      return positions.filter(p => p.platform === platform);
    }
    
    return positions;
  }

  /**
   * Get single position details
   * @param {string} positionId - Position identifier
   */
  getPosition(positionId) {
    return this.positions.get(positionId);
  }

  /**
   * Set stop-loss for a position
   * @param {Object} params - Stop-loss parameters
   * @param {string} params.positionId - Position ID
   * @param {number} params.price - Stop price (absolute)
   * @param {number} params.percentFromEntry - Stop percentage from entry
   * @param {number} params.sizePercent - Percentage of position to exit
   */
  async setStopLoss({ positionId, price = null, percentFromEntry = null, sizePercent = 100 }) {
    const pos = this.positions.get(positionId);
    if (!pos) throw new Error(`Position ${positionId} not found`);
    
    let stopPrice = price;
    
    if (percentFromEntry && !price) {
      // Calculate percentage from entry
      const direction = pos.side === 'YES' || pos.side === 'long' ? 1 : -1;
      stopPrice = pos.entryPrice * (1 - (percentFromEntry / 100) * direction);
    }
    
    pos.stopLoss = {
      price: stopPrice,
      sizePercent,
      setAt: new Date()
    };
    
    this.positions.set(positionId, pos);
    this.emit('stopLossSet', { positionId, stopPrice, sizePercent });
    
    return { positionId, stopPrice, sizePercent };
  }

  /**
   * Set take-profit for a position
   * @param {Object} params - Take-profit parameters
   * @param {string} params.positionId - Position ID
   * @param {number} params.price - Target price (absolute)
   * @param {number} params.percentFromEntry - Target percentage from entry
   * @param {Array} params.levels - Multiple take-profit levels
   */
  async setTakeProfit({ positionId, price = null, percentFromEntry = null, levels = null }) {
    const pos = this.positions.get(positionId);
    if (!pos) throw new Error(`Position ${positionId} not found`);
    
    if (levels) {
      pos.takeProfit = {
        type: 'multi-level',
        levels: levels.map(level => ({
          price: level.price,
          sizePercent: level.sizePercent,
          triggered: false
        })),
        setAt: new Date()
      };
    } else {
      let targetPrice = price;
      
      if (percentFromEntry && !price) {
        const direction = pos.side === 'YES' || pos.side === 'long' ? 1 : -1;
        targetPrice = pos.entryPrice * (1 + (percentFromEntry / 100) * direction);
      }
      
      pos.takeProfit = {
        type: 'single',
        price: targetPrice,
        sizePercent: 100,
        setAt: new Date()
      };
    }
    
    this.positions.set(positionId, pos);
    this.emit('takeProfitSet', { positionId, takeProfit: pos.takeProfit });
    
    return { positionId, takeProfit: pos.takeProfit };
  }

  /**
   * Set trailing stop for a position
   * @param {Object} params - Trailing stop parameters
   * @param {string} params.positionId - Position ID
   * @param {number} params.trailPercent - Percentage to trail from high
   * @param {number} params.activateAt - Price to activate trailing stop
   */
  async setTrailingStop({ positionId, trailPercent, activateAt = null }) {
    const pos = this.positions.get(positionId);
    if (!pos) throw new Error(`Position ${positionId} not found`);
    
    pos.trailingStop = {
      trailPercent,
      activateAt,
      isActive: activateAt ? false : true,
      setAt: new Date()
    };
    
    this.positions.set(positionId, pos);
    this.emit('trailingStopSet', { positionId, trailPercent, activateAt });
    
    return { positionId, trailPercent, activateAt };
  }

  /**
   * Remove all stops from a position
   * @param {string} positionId - Position identifier
   */
  async removeAllStops(positionId) {
    const pos = this.positions.get(positionId);
    if (!pos) throw new Error(`Position ${positionId} not found`);
    
    pos.stopLoss = null;
    pos.takeProfit = null;
    pos.trailingStop = null;
    
    this.positions.set(positionId, pos);
    this.emit('stopsRemoved', positionId);
    
    return { positionId, removed: true };
  }

  /**
   * Monitor positions and trigger stops
   */
  async monitorPositions() {
    for (const [id, pos] of this.positions) {
      await this.checkStopLoss(pos);
      await this.checkTakeProfit(pos);
      await this.checkTrailingStop(pos);
    }
  }

  async checkStopLoss(pos) {
    if (!pos.stopLoss) return;
    
    const shouldTrigger = pos.side === 'YES' || pos.side === 'long'
      ? pos.currentPrice <= pos.stopLoss.price
      : pos.currentPrice >= pos.stopLoss.price;
    
    if (shouldTrigger) {
      this.emit('stopLossTriggered', {
        position: pos,
        exitPrice: pos.currentPrice,
        pnl: this.calculatePnL(pos),
        sizePercent: pos.stopLoss.sizePercent
      });
      
      // Close or reduce position
      if (pos.stopLoss.sizePercent >= 100) {
        this.positions.delete(pos.id);
      } else {
        // Partial exit - reduce size
        pos.size = pos.size * (1 - pos.stopLoss.sizePercent / 100);
        pos.stopLoss = null;
        this.positions.set(pos.id, pos);
      }
    }
  }

  async checkTakeProfit(pos) {
    if (!pos.takeProfit) return;
    
    if (pos.takeProfit.type === 'single') {
      const shouldTrigger = pos.side === 'YES' || pos.side === 'long'
        ? pos.currentPrice >= pos.takeProfit.price
        : pos.currentPrice <= pos.takeProfit.price;
      
      if (shouldTrigger) {
        this.emit('takeProfitTriggered', {
          position: pos,
          exitPrice: pos.currentPrice,
          pnl: this.calculatePnL(pos)
        });
        this.positions.delete(pos.id);
      }
    } else if (pos.takeProfit.type === 'multi-level') {
      for (const level of pos.takeProfit.levels) {
        if (!level.triggered) {
          const shouldTrigger = pos.side === 'YES' || pos.side === 'long'
            ? pos.currentPrice >= level.price
            : pos.currentPrice <= level.price;
          
          if (shouldTrigger) {
            level.triggered = true;
            this.emit('takeProfitTriggered', {
              position: pos,
              exitPrice: pos.currentPrice,
              sizePercent: level.sizePercent,
              level: level
            });
            
            // Partial exit
            pos.size = pos.size * (1 - level.sizePercent / 100);
            this.positions.set(pos.id, pos);
          }
        }
      }
      
      // Check if all levels triggered
      const allTriggered = pos.takeProfit.levels.every(l => l.triggered);
      if (allTriggered) {
        this.positions.delete(pos.id);
      }
    }
  }

  async checkTrailingStop(pos) {
    if (!pos.trailingStop) return;
    
    const trail = pos.trailingStop;
    
    // Check activation condition
    if (!trail.isActive && trail.activateAt) {
      if (pos.currentPrice >= trail.activateAt) {
        trail.isActive = true;
        pos.highWaterMark = pos.currentPrice;
        this.emit('trailingStopActivated', { positionId: pos.id });
      } else {
        return;
      }
    }
    
    if (!trail.isActive) return;
    
    const trailPrice = pos.highWaterMark * (1 - trail.trailPercent / 100);
    const shouldTrigger = pos.currentPrice <= trailPrice;
    
    if (shouldTrigger) {
      this.emit('trailingStopTriggered', {
        position: pos,
        exitPrice: pos.currentPrice,
        highWaterMark: pos.highWaterMark,
        pnl: this.calculatePnL(pos)
      });
      this.positions.delete(pos.id);
    }
  }

  calculatePnL(position) {
    const priceDiff = position.currentPrice - position.entryPrice;
    const direction = position.side === 'YES' || position.side === 'long' ? 1 : -1;
    const pnl = direction * priceDiff * position.size;
    const pnlPercent = (pnl / (position.entryPrice * position.size)) * 100;
    
    return { pnl, pnlPercent };
  }

  /**
   * Get position summary
   */
  async getSummary() {
    const positions = Array.from(this.positions.values());
    let totalValue = 0;
    let totalPnl = 0;
    let withStopLoss = 0;
    let withTakeProfit = 0;
    
    for (const pos of positions) {
      const value = pos.currentPrice * pos.size;
      totalValue += value;
      totalPnl += this.calculatePnL(pos).pnl;
      if (pos.stopLoss) withStopLoss++;
      if (pos.takeProfit) withTakeProfit++;
    }
    
    return {
      count: positions.length,
      totalValue,
      unrealizedPnl: totalPnl,
      withStopLoss,
      withTakeProfit,
      withTrailingStop: positions.filter(p => p.trailingStop).length
    };
  }

  generateId() {
    return `pos_${Date.now()}_${Math.random().toString(36).substr(2, 8)}`;
  }
}

export default PositionManager;

File: position_manager.go
package main

import (
	"fmt"
	"math/rand"
	"sync"
	"time"
)

type StopLoss struct {
	Price       float64   `json:"price"`
	SizePercent float64   `json:"sizePercent"`
	SetAt       time.Time `json:"setAt"`
}

type TakeProfit struct {
	Type        string            `json:"type"`
	Price       float64           `json:"price,omitempty"`
	SizePercent float64           `json:"sizePercent,omitempty"`
	Levels      []TakeProfitLevel `json:"levels,omitempty"`
	SetAt       time.Time         `json:"setAt"`
}

type TakeProfitLevel struct {
	Price       float64 `json:"price"`
	SizePercent float64 `json:"sizePercent"`
	Triggered   bool    `json:"triggered"`
}

type TrailingStop struct {
	TrailPercent float64   `json:"trailPercent"`
	ActivateAt   float64   `json:"activateAt,omitempty"`
	IsActive     bool      `json:"isActive"`
	SetAt        time.Time `json:"setAt"`
}

type Position struct {
	ID            string        `json:"id"`
	Platform      string        `json:"platform"`
	Market        string        `json:"market"`
	Side          string        `json:"side"`
	Size          float64       `json:"size"`
	EntryPrice    float64       `json:"entryPrice"`
	CurrentPrice  float64       `json:"currentPrice"`
	StopLoss      *StopLoss     `json:"stopLoss"`
	TakeProfit    *TakeProfit   `json:"takeProfit"`
	TrailingStop  *TrailingStop `json:"trailingStop"`
	HighWaterMark float64       `json:"highWaterMark"`
	CreatedAt     time.Time     `json:"createdAt"`
	UpdatedAt     time.Time     `json:"updatedAt"`
}

type PnL struct {
	PnL        float64 `json:"pnl"`
	PnLPercent float64 `json:"pnlPercent"`
}

type PositionEvent struct {
	Type     string
	Position Position
	PnL      *PnL
	Extra    map[string]interface{}
}

type PositionSummary struct {
	Count             int     `json:"count"`
	TotalValue        float64 `json:"totalValue"`
	UnrealizedPnl     float64 `json:"unrealizedPnl"`
	WithStopLoss      int     `json:"withStopLoss"`
	WithTakeProfit    int     `json:"withTakeProfit"`
	WithTrailingStop  int     `json:"withTrailingStop"`
}

type PositionManager struct {
	mu             sync.RWMutex
	positions      map[string]*Position
	checkInterval  time.Duration
	ticker         *time.Ticker
	done           chan struct{}
	EventHandlers  map[string][]func(PositionEvent)
}

func NewPositionManager(checkIntervalMs int) *PositionManager {
	if checkIntervalMs <= 0 {
		checkIntervalMs = 5000
	}
	return &PositionManager{
		positions:     make(map[string]*Position),
		checkInterval: time.Duration(checkIntervalMs) * time.Millisecond,
		done:          make(chan struct{}),
		EventHandlers: make(map[string][]func(PositionEvent)),
	}
}

func (pm *PositionManager) On(event string, handler func(PositionEvent)) {
	pm.EventHandlers[event] = append(pm.EventHandlers[event], handler)
}

func (pm *PositionManager) emit(event string, e PositionEvent) {
	for _, h := range pm.EventHandlers[event] {
		go h(e)
	}
}

func (pm *PositionManager) Start() {
	pm.ticker = time.NewTicker(pm.checkInterval)
	go func() {
		for {
			select {
			case <-pm.ticker.C:
				pm.monitorPositions()
			case <-pm.done:
				return
			}
		}
	}()
	fmt.Println("Position manager started")
}

func (pm *PositionManager) Stop() {
	if pm.ticker != nil {
		pm.ticker.Stop()
	}
	close(pm.done)
}

func (pm *PositionManager) AddPosition(platform, market, side string, size, entryPrice float64) *Position {
	pos := &Position{
		ID:            generatePositionID(),
		Platform:      platform,
		Market:        market,
		Side:          side,
		Size:          size,
		EntryPrice:    entryPrice,
		CurrentPrice:  entryPrice,
		HighWaterMark: entryPrice,
		CreatedAt:     time.Now(),
		UpdatedAt:     time.Now(),
	}

	pm.mu.Lock()
	pm.positions[pos.ID] = pos
	pm.mu.Unlock()

	pm.emit("positionAdded", PositionEvent{Type: "positionAdded", Position: *pos})
	return pos
}

func (pm *PositionManager) UpdatePrice(positionID string, currentPrice float64) *Position {
	pm.mu.Lock()
	defer pm.mu.Unlock()

	pos, ok := pm.positions[positionID]
	if !ok {
		return nil
	}
	pos.CurrentPrice = currentPrice
	pos.UpdatedAt = time.Now()
	if currentPrice > pos.HighWaterMark {
		pos.HighWaterMark = currentPrice
	}
	return pos
}

func (pm *PositionManager) ListPositions(platform string) []Position {
	pm.mu.RLock()
	defer pm.mu.RUnlock()

	result := make([]Position, 0)
	for _, p := range pm.positions {
		if platform == "" || p.Platform == platform {
			result = append(result, *p)
		}
	}
	return result
}

func (pm *PositionManager) GetPosition(positionID string) *Position {
	pm.mu.RLock()
	defer pm.mu.RUnlock()
	if p, ok := pm.positions[positionID]; ok {
		cp := *p
		return &cp
	}
	return nil
}

func (pm *PositionManager) SetStopLoss(positionID string, price, percentFromEntry, sizePercent float64) error {
	pm.mu.Lock()
	defer pm.mu.Unlock()

	pos, ok := pm.positions[positionID]
	if !ok {
		return fmt.Errorf("position %s not found", positionID)
	}

	stopPrice := price
	if percentFromEntry > 0 && price == 0 {
		direction := 1.0
		if pos.Side != "YES" && pos.Side != "long" {
			direction = -1
		}
		stopPrice = pos.EntryPrice * (1 - (percentFromEntry/100)*direction)
	}
	if sizePercent == 0 {
		sizePercent = 100
	}

	pos.StopLoss = &StopLoss{Price: stopPrice, SizePercent: sizePercent, SetAt: time.Now()}
	return nil
}

func (pm *PositionManager) SetTakeProfit(positionID string, price, percentFromEntry float64, levels []TakeProfitLevel) error {
	pm.mu.Lock()
	defer pm.mu.Unlock()

	pos, ok := pm.positions[positionID]
	if !ok {
		return fmt.Errorf("position %s not found", positionID)
	}

	if len(levels) > 0 {
		pos.TakeProfit = &TakeProfit{Type: "multi-level", Levels: levels, SetAt: time.Now()}
	} else {
		targetPrice := price
		if percentFromEntry > 0 && price == 0 {
			direction := 1.0
			if pos.Side != "YES" && pos.Side != "long" {
				direction = -1
			}
			targetPrice = pos.EntryPrice * (1 + (percentFromEntry/100)*direction)
		}
		pos.TakeProfit = &TakeProfit{Type: "single", Price: targetPrice, SizePercent: 100, SetAt: time.Now()}
	}
	return nil
}

func (pm *PositionManager) SetTrailingStop(positionID string, trailPercent, activateAt float64) error {
	pm.mu.Lock()
	defer pm.mu.Unlock()

	pos, ok := pm.positions[positionID]
	if !ok {
		return fmt.Errorf("position %s not found", positionID)
	}

	pos.TrailingStop = &TrailingStop{
		TrailPercent: trailPercent,
		ActivateAt:   activateAt,
		IsActive:     activateAt == 0,
		SetAt:        time.Now(),
	}
	return nil
}

func (pm *PositionManager) RemoveAllStops(positionID string) error {
	pm.mu.Lock()
	defer pm.mu.Unlock()

	pos, ok := pm.positions[positionID]
	if !ok {
		return fmt.Errorf("position %s not found", positionID)
	}
	pos.StopLoss = nil
	pos.TakeProfit = nil
	pos.TrailingStop = nil
	return nil
}

func (pm *PositionManager) DeletePosition(positionID string) {
	pm.mu.Lock()
	delete(pm.positions, positionID)
	pm.mu.Unlock()
}

func (pm *PositionManager) GetSummary() PositionSummary {
	pm.mu.RLock()
	defer pm.mu.RUnlock()

	s := PositionSummary{}
	for _, pos := range pm.positions {
		s.Count++
		s.TotalValue += pos.CurrentPrice * pos.Size
		pnl := pm.calculatePnL(pos)
		s.UnrealizedPnl += pnl.PnL
		if pos.StopLoss != nil {
			s.WithStopLoss++
		}
		if pos.TakeProfit != nil {
			s.WithTakeProfit++
		}
		if pos.TrailingStop != nil {
			s.WithTrailingStop++
		}
	}
	return s
}

func (pm *PositionManager) monitorPositions() {
	pm.mu.Lock()
	ids := make([]string, 0, len(pm.positions))
	for id := range pm.positions {
		ids = append(ids, id)
	}
	pm.mu.Unlock()

	for _, id := range ids {
		pm.mu.RLock()
		pos, ok := pm.positions[id]
		if !ok {
			pm.mu.RUnlock()
			continue
		}
		posCopy := *pos
		pm.mu.RUnlock()

		pm.checkStopLoss(&posCopy)
		pm.checkTakeProfit(&posCopy)
		pm.checkTrailingStop(&posCopy)
	}
}

func (pm *PositionManager) checkStopLoss(pos *Position) {
	if pos.StopLoss == nil {
		return
	}
	isLong := pos.Side == "YES" || pos.Side == "long"
	shouldTrigger := (isLong && pos.CurrentPrice <= pos.StopLoss.Price) ||
		(!isLong && pos.CurrentPrice >= pos.StopLoss.Price)

	if shouldTrigger {
		pnl := pm.calculatePnL(pos)
		pm.emit("stopLossTriggered", PositionEvent{
			Type:     "stopLossTriggered",
			Position: *pos,
			PnL:      &pnl,
		})
		if pos.StopLoss.SizePercent >= 100 {
			pm.mu.Lock()
			delete(pm.positions, pos.ID)
			pm.mu.Unlock()
		} else {
			pm.mu.Lock()
			if p, ok := pm.positions[pos.ID]; ok {
				p.Size = p.Size * (1 - p.StopLoss.SizePercent/100)
				p.StopLoss = nil
			}
			pm.mu.Unlock()
		}
	}
}

func (pm *PositionManager) checkTakeProfit(pos *Position) {
	if pos.TakeProfit == nil {
		return
	}
	isLong := pos.Side == "YES" || pos.Side == "long"

	if pos.TakeProfit.Type == "single" {
		shouldTrigger := (isLong && pos.CurrentPrice >= pos.TakeProfit.Price) ||
			(!isLong && pos.CurrentPrice <= pos.TakeProfit.Price)
		if shouldTrigger {
			pnl := pm.calculatePnL(pos)
			pm.emit("takeProfitTriggered", PositionEvent{Type: "takeProfitTriggered", Position: *pos, PnL: &pnl})
			pm.mu.Lock()
			delete(pm.positions, pos.ID)
			pm.mu.Unlock()
		}
	} else if pos.TakeProfit.Type == "multi-level" {
		allTriggered := true
		for i, level := range pos.TakeProfit.Levels {
			if !level.Triggered {
				shouldTrigger := (isLong && pos.CurrentPrice >= level.Price) ||
					(!isLong && pos.CurrentPrice <= level.Price)
				if shouldTrigger {
					pm.mu.Lock()
					if p, ok := pm.positions[pos.ID]; ok {
						p.TakeProfit.Levels[i].Triggered = true
						p.Size = p.Size * (1 - level.SizePercent/100)
					}
					pm.mu.Unlock()
					pnl := pm.calculatePnL(pos)
					pm.emit("takeProfitTriggered", PositionEvent{Type: "takeProfitTriggered", Position: *pos, PnL: &pnl})
				} else {
					allTriggered = false
				}
			}
		}
		if allTriggered {
			pm.mu.Lock()
			delete(pm.positions, pos.ID)
			pm.mu.Unlock()
		}
	}
}

func (pm *PositionManager) checkTrailingStop(pos *Position) {
	if pos.TrailingStop == nil {
		return
	}
	trail := pos.TrailingStop

	if !trail.IsActive && trail.ActivateAt > 0 {
		if pos.CurrentPrice >= trail.ActivateAt {
			pm.mu.Lock()
			if p, ok := pm.positions[pos.ID]; ok {
				p.TrailingStop.IsActive = true
				p.HighWaterMark = pos.CurrentPrice
			}
			pm.mu.Unlock()
		} else {
			return
		}
	}

	if !trail.IsActive {
		return
	}

	trailPrice := pos.HighWaterMark * (1 - trail.TrailPercent/100)
	if pos.CurrentPrice <= trailPrice {
		pnl := pm.calculatePnL(pos)
		pm.emit("trailingStopTriggered", PositionEvent{Type: "trailingStopTriggered", Position: *pos, PnL: &pnl})
		pm.mu.Lock()
		delete(pm.positions, pos.ID)
		pm.mu.Unlock()
	}
}

func (pm *PositionManager) calculatePnL(pos *Position) PnL {
	priceDiff := pos.CurrentPrice - pos.EntryPrice
	direction := 1.0
	if pos.Side != "YES" && pos.Side != "long" {
		direction = -1
	}
	pnl := direction * priceDiff * pos.Size
	pnlPercent := (pnl / (pos.EntryPrice * pos.Size)) * 100
	return PnL{PnL: pnl, PnLPercent: pnlPercent}
}

func generatePositionID() string {
	return fmt.Sprintf("pos_%d_%s", time.Now().UnixMilli(), randStr(8))
}

func randStr(n int) string {
	const letters = "abcdefghijklmnopqrstuvwxyz0123456789"
	b := make([]byte, n)
	for i := range b {
		b[i] = letters[rand.Intn(len(letters))]
	}
	return string(b)
}
